### ISSUE TITLE
latest XGBoost Python native API `custom_metric` vs `disable_default_eval_metric` not helping learning.

### ISSUE THREAD

**Issue by pangjac** [dmlc/xgboost#9851]
Dear community,

I'm using the latest version of XGBoost (2.0.2) and have found that the Python API's `disable_default_eval_metric` and `custom_eval parameters` don't seem to affect learning as expected:
- I've created a customized metric function named `eval_metric_accuracy_customized`. This function focuses on measuring the accuracy of predictions specifically for target values (`y_true`) that fall within the 15% to 85% range of all y_true values. In other words, it prioritizes prediction accuracy within this range and is less concerned with outliers. The goal is to **maximize** this metric, as a higher value indicates better performance for the predictions that matter most in this specified range. (Reference: [XGBoost Python API](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.train))

The conflicted observations are :

- The `maximize` parameter in the train API doesn't seem to affect the training: whether set to `True` or `False`, the output for eval_metric_accuracy_xgboost during each boosting round remains the same. 
  - I was expecting by setting `maximize` True, the printed out of each num of round should be observed improved (i.e, slightly increasing `eval_metric_accuracy_xgboost` value for each round, but this metric for both train/validation oscillate without much improvement), and setting `maximize` to False does not impact the learning too. 
  - **See below comparison Test 1 and Test 2.**

- The `disable_default_eval_metric` parameter's only effect is whether on the display of the `RMSE` metric. Changing this setting doesn't alter the training process or the final prediction accuracy. 
  - I was expecting by setting `disable_default_eval_metric` True, the xgboost learning procedure should focus on customized metric, while if False, then the xgboost learning procedure should focus on default metric (in this case, `rmse`), such that the printed out of each num of round should be observed somewhat difference, but there is not. 
  - **See below comparison Test 1 and Test 3.**

When I combine `disable_default_eval_metric` with my custom metric `eval_metric_accuracy_xgboost` and toggle `maximize` between `True` and `False`, **the four of experiments yield identical prediction scores**(?!)

Below is the code example. And if you can access Google Colab, **here is the link for Google colab to replicate the result** : https://colab.research.google.com/drive/1KxzOT25AVUgcDW6GubvmjFsytfhJQeh0?usp=sharing

```
# Load the data
housing = fetch_california_housing()
X_train, X_test, y_train, y_test = train_test_split(
    pd.DataFrame(housing.data, columns=housing.feature_names),
    housing.target,
    test_size=0.25,
    random_state=131,
)

import xgboost as xgb
from typing import List, Tuple

# data
dtrain = xgb.DMatrix(data=X_train, label=y_train, missing=-999999999)
dvalid = xgb.DMatrix(data=X_test, label=y_test, missing=-999999999)

# made up xgboost params 
param = {'objective': 'reg:squarederror', 
           'disable_default_eval_metric': True, # <---------Noticed disable_default_eval_metric is True 
          'tree_method': 'hist',
          'booster': 'dart', 
          'lambda': 0.5022935723779454, 'alpha': 0.0010591193559734626, 
          'subsample': 0.7443155004860621, 'colsample_bytree': 0.8049514766470095, 
          'base_score': 2.0729023397932815, 
          'eta': 1.146933698699281, 'gamma': 3.1491135525631537, 'max_depth': 8, 'min_child_weight': 60, 
          'grow_policy': 'lossguide', 'sample_type': 'uniform', 
          'normalize_type': 'forest', 'rate_drop': 0.0001559101552440383, 'skip_drop': 0.006540833684514242}

# customizd metric
LB = np.percentile(y_train, 15)
UB = np.percentile(y_train, 85)
def eval_metric_accuracy_customized(y_pred, y_true, sample_weight=None, LB=0, UB=np.inf, threshold=0.25):
    # cond checks if y_pred are within the tolerance range for y_true which lies between [LB, UB]
    cond = (y_pred >= (1-threshold)*y_true) & \
            (y_pred <= (1+threshold)*y_true) & \
            (y_true > LB) & (y_true < UB)
    df_true_score_weight_1 = np.where(cond, 1.0, 0.0)
    acc = np.mean(df_true_score_weight_1)
    # acc indicate the prportion of y_pred fall within the condition of cond
    return acc
def eval_metric_accuracy_xgboost(sample_weight=None, LB=0, UB=np.inf, threshold=0.25,  **kwargs):
    def eval_metric_accuracy_xgboost_internal(predt: np.ndarray, dtrain: xgb.DMatrix) -> Tuple[str, float]:
        y_true = dtrain.get_label()
        score = eval_metric_accuracy_customized(y_pred=predt, y_true=y_true, 
                                                sample_weight=sample_weight, LB=LB, UB=UB, threshold=threshold)
        return "eval_metric_accuracy_xgboost", score
    return eval_metric_accuracy_xgboost_internal
eval_metric_accuracy_xgboost_custom = eval_metric_accuracy_xgboost(LB=LB, UB=UB)

# Test 1
# parame disable_default_eval_metric true
# customized metric accuracy set to `maximize` because we want xgboost to optimize acc as much as possible
output = xgb.train(params=param, 
                   dtrain=dtrain, 
                   num_boost_round=200,  # set a high number
                   evals=[(dtrain, "train"),(dvalid, "validation")], 
                   custom_metric=eval_metric_accuracy_xgboost_custom,
                   maximize=True,   # <---------Noticed   `custom_metric` specified and metric direction maximize True
                   verbose_eval=True
                   )

preds = output.predict(data=dvalid)
print("Test RMSE:", mean_squared_error(y_test, preds, squared=False)) # Test RMSE: 0.5912299076469258


# Test 2
# parame disable_default_eval_metric true
# customized metric accuracy set to NOT `maximize`
output_maximize_accuracy_false = xgb.train(params=param, 
                   dtrain=dtrain, 
                   num_boost_round=200,
                   evals=[(dtrain, "train"),(dvalid, "validation")], 
                   feval=eval_metric_accuracy_xgboost_custom,
                   maximize=False, # <---------Noticed   `custom_metric` specified and metric direction maximize False
                   verbose_eval=True
                   )

preds_output_maximize_accuracy_false = output_maximize_accuracy_false.predict(data=dvalid)
print("Test RMSE:", mean_squared_error(y_test, preds_output_maximize_accuracy_false, squared=False)) # Test RMSE: 0.5912299076469258

# Test 3
# parame disable_default_eval_metric false
# customized metric accuracy set to `maximize`
param_disablt_default_eval_false = {'objective': 'reg:squarederror', 
                                    'disable_default_eval_metric': False,  # <---------Noticed disable_default_eval_metric is True 
                                    'tree_method': 'hist',
                                    'booster': 'dart', 
                                    'lambda': 0.5022935723779454, 'alpha': 0.0010591193559734626, 
                                    'subsample': 0.7443155004860621, 'colsample_bytree': 0.8049514766470095, 
                                    'base_score': 2.0729023397932815, 
                                    'eta': 1.146933698699281, 'gamma': 3.1491135525631537, 'max_depth': 8, 'min_child_weight': 60, 
                                    'grow_policy': 'lossguide', 'sample_type': 'uniform', 
                                    'normalize_type': 'forest', 'rate_drop': 0.0001559101552440383, 'skip_drop': 0.006540833684514242}

output_maximize_accuracy_true_disable_default_false = xgb.train(params=param_disablt_default_eval_false, 
                   dtrain=dtrain, 
                   num_boost_round=200,
                   evals=[(dtrain, "train"),(dvalid, "validation")], 
                   feval=eval_metric_accuracy_xgboost_custom,
                   maximize=True,  # <---------Noticed   `custom_metric` specified and metric direction maximize True
                   verbose_eval=True
                   )

preds_output_maximize_accuracy_true_disable_default_false = output_maximize_accuracy_true_disable_default_false.predict(data=dvalid)
print("Test RMSE:", mean_squared_error(y_test, preds_output_maximize_accuracy_true_disable_default_false, squared=False)) # Test RMSE: 0.5912299076469258

# Test 4 
# parame disable_default_eval_metric false
# customized metric accuracy set to NOT `maximize`
output_maximize_accuracy_false_disable_default_false = xgb.train(params=param_disablt_default_eval_false, 
                   dtrain=dtrain, 
                   num_boost_round=200, 
                   evals=[(dtrain, "train"),(dvalid, "validation")], 
                   feval=eval_metric_accuracy_xgboost_custom,
                   maximize=False,  # <---------Noticed   `custom_metric` specified and metric direction maximize False
                   verbose_eval=True
                   )

preds_output_maximize_accuracy_false_disable_default_false = output_maximize_accuracy_false_disable_default_false.predict(data=dvalid)
print("Test RMSE:", mean_squared_error(y_test, preds_output_maximize_accuracy_false_disable_default_false, squared=False)) # Test RMSE: 0.5912299076469258
```

I have also raised the question in discussion forum https://discuss.xgboost.ai/t/replicate-code-notebook-attached-xgboost-custom-metric-vis-disable-default-eval-metric-not-working/3455

**Comment by trivialfis** [comment]
Let me take a look tomorrow.

**Comment by trivialfis** [comment]
Feel free to reopen if there's further question.
